idioms/coercion-arguments.md

```diff
@@ -0,0 +1,117 @@
+# Use borrowed types for arguments
+
+## Description
+
+Using a target of a deref coercion can increase the flexibility of your code when you are deciding which argument type to use for a function argument.
+In this way, the function will accept more input types.
+
+For example; using `&str` instead of a `&String`, or `&[T]` in preference of `&Vec<T>`.  This is not limited to slice-able or fat pointer types. In fact you should always prefer using the __borrowed type__ over __borrowing the owned type__, e.g., `&T` to `&Box<T>`.
+
+Using borrowed types you can avoid layers of indirection for those instances where the owned type already provides a layer of indirection, as it will be illustrated in the following examples. For instance, a `String` has a layer of indirection, so a `&String` will have two layers of indrection.
+We can avoid this by using `&str` instead, and letting `&String` coerce to a `&str` whenever the function is invoked. 
+
+## Example
+
+For this example, we will illustrate some differences for using `&String` as a function argument versus using a `&str`, but the ideas apply as well to using `&Vec<T>` versus using a `&[T]` or using a `&T` versus a `&Box<T>`.
+
+Consider an example where we wish to determine if a word contains three consecutive vowels.
+We don't need to own the string to determine this, so we will take a reference.
+
+The code might look something like this:
+
+```rust
+fn three_vowels(word: &String) -> bool {
+    let mut vowel_count = 0;
+    for c in word.chars() {
+        match c {
+            'a' | 'e' | 'i' | 'o' | 'u' => {
+                vowel_count += 1;
+                if vowel_count >= 3 {
+                    return true
+                }
+            }
+            _ => vowel_count = 0
+        }
+    }
+    false
```

Comment:
Would using bytes here be a better choice?

Reply:
bytes instead of characters? Why? It might look scary for beginners.

Reply:
Is it really that scary, it's just low-effort change from `'a'` to `b'a'` and `chars()` to `bytes()` for a lot more performance?

Reply:
That would be good if you would do that in your own code when using a `three_vowels` function that is optimized for performance, but I think for examples it's more about what they show overall, and this example is not about performance or `chars vs bytes` but about `&str vs &String`  or Coercion in function arguments. So as I know what you mean, I think changing this to bytes is not overly relevant to this example.

Reply:
Agreed. Using bytes introduces another concept (UTF-8 encoding) which distracts from the main point.